4_basic_assembly/2_bitwise_operations/2_practical_bit_games/answers/3_write_code/2.asm

```diff
@@ -0,0 +1,51 @@
+; Basic Assembly
+; ==============
+;
+; 2.  Bit reverse:
+;
+;     Write a program that takes a number (of size 4 bytes) x as input, and then
+;     reverses all the bits of x, and outputs the result. By reversing all bits we
+;     mean that the bit with original location i will move to location 31-i.
+; 
+;     Small example (for the 8 bit case):
+;
+;       if x == {01001111}_2, then the output is {11110010}_2.
+;
+;       In this example we reversed only 8 bits. Your program will be able to
+;       reverse 32 bits.
+;
+; More ways to reverse bits can be found here:
+; http://graphics.stanford.edu/~seander/bithacks.html#BitReverseObvious
+
+format PE console
+entry start
+
+include 'win32a.inc' 
+
+; ===============================================
+section '.text' code readable executable
+
+start:
+
+    call    read_hex
+    ; mov     eax,01001111b
+    mov     ecx,31d
+    xor     ebx,ebx
+    xor     esi,esi
+
+rotate:
+    mov     ebx,1
+    and     ebx,eax
+    add     esi,ebx
+    ror     eax,1
+    shl     esi,1

```

Comment:
@liara : I didn't run this code, but I think that it will not always return a correct result. This is because the last thing we do with esi is shifting left. As a result, we can know for sure that the lowest bit of esi in the end of this loop is always 0.

If for example we began from a number that has 1 as the most significant bit, how will this be reflected in this code? For example, if we begin with 0xfffffff0, we expect to get 0x0fffffff. Could you check if this code works with this input?

All that said, obviously you got the idea of how to bit reverse right. You unroll one bit every time from eax, and feed it back into esi. 
I don't remember if the course covers it, but there is an instruction called RCR which allows you to rotate while feeding in the last carry flag. It could be very useful in this code, as the SHR instruction sets the carry flag to be the last bit that has disappeared.
I also suspect that something is

Reply:
Yeah, as you said, inputing 0xfffffff0  gives 0xfffffffe instead of 0x0fffffff. But there were a lot of correct outputs otherwise:P
I tested the solution with RCR, but It was not mentioned in the course, so I thought I'm not supposed to use it. I guess I can then use RCL too?